## House Prices - Advanced Regression Techniques

## notes
target = SalePrice  

problems: too many features...

In [1]:
import pandas as pd

In [2]:
df = pd.read_csv('./train.csv')

In [3]:
df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [4]:
df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 81 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1460 non-null   int64  
 1   MSSubClass     1460 non-null   int64  
 2   MSZoning       1460 non-null   object 
 3   LotFrontage    1201 non-null   float64
 4   LotArea        1460 non-null   int64  
 5   Street         1460 non-null   object 
 6   Alley          91 non-null     object 
 7   LotShape       1460 non-null   object 
 8   LandContour    1460 non-null   object 
 9   Utilities      1460 non-null   object 
 10  LotConfig      1460 non-null   object 
 11  LandSlope      1460 non-null   object 
 12  Neighborhood   1460 non-null   object 
 13  Condition1     1460 non-null   object 
 14  Condition2     1460 non-null   object 
 15  BldgType       1460 non-null   object 
 16  HouseStyle     1460 non-null   object 
 17  OverallQual    1460 non-null   int64  
 18  OverallC

## Handling the missing values

In [6]:
print(df.isnull().sum().sort_values(ascending=False).head(20))

PoolQC          1453
MiscFeature     1406
Alley           1369
Fence           1179
MasVnrType       872
FireplaceQu      690
LotFrontage      259
GarageYrBlt       81
GarageCond        81
GarageType        81
GarageFinish      81
GarageQual        81
BsmtFinType2      38
BsmtExposure      38
BsmtQual          37
BsmtCond          37
BsmtFinType1      37
MasVnrArea         8
Electrical         1
Id                 0
dtype: int64


In [7]:
print(f'{df.isnull().sum().sort_values(ascending=False).head(20).sum()} missing values')
cat_features = df.select_dtypes(include=['object']).shape[1]
print(f"{cat_features} categorical features")
num_features = df.select_dtypes(include=['int64', 'float64']).shape[1]
print(f"{num_features} numerical features")

7829 missing values
43 categorical features
38 numerical features


In [8]:
df.shape

(1460, 81)

## Reports
We have 1460 enteries and 81 features 
7829 missing values  
43 categorical features  
38 numerical features  
whats scary? too many features and too many missing values.

## Training a linear model, without the missing values

In [9]:
# Select only numerical columns
df_numeric = df.select_dtypes(include=['int64', 'float64'])

# Drop columns with any missing values
df_clean = df_numeric.dropna(axis=0)

print(f"Original shape: {df.shape}")
print(f"Numeric shape: {df_numeric.shape}")
print(f"Clean shape: {df_clean.shape}")

Original shape: (1460, 81)
Numeric shape: (1460, 38)
Clean shape: (1121, 38)


## Spliting the data

In [10]:
from sklearn.model_selection import train_test_split, cross_val_score

In [11]:
X = df_clean.drop(['SalePrice', 'Id'], axis=1)
y = df_clean['SalePrice']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

## Training the model

In [12]:
from sklearn.linear_model import LinearRegression

In [13]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [14]:
coef_df0 = pd.DataFrame({
    'Feature': X.columns,
    'Coefficient': model.coef_
})
coef_df0 = coef_df0.sort_values('Coefficient', ascending=False).reset_index(drop=True)
coef_df0

Feature   Coefficient
0     OverallQual  19654.291953
1      GarageCars  18344.552805
2    BsmtFullBath  11565.128964
3     OverallCond   5072.289462
4        FullBath   5010.143758
5      Fireplaces   4793.918236
6    TotRmsAbvGrd   4587.157867
7    BsmtHalfBath   1338.748402
8       YearBuilt    286.898275
9    YearRemodAdd    125.155507
10    ScreenPorch     66.973618
11      3SsnPorch     58.657103
12      GrLivArea     29.344522
13     MasVnrArea     26.492720
14     WoodDeckSF     20.256803
15       2ndFlrSF     13.903081
16  EnclosedPorch     11.847823
17       1stFlrSF     10.350316
18    TotalBsmtSF      6.237704
19     GarageArea      5.860430
20   LowQualFinSF      5.091125
21     BsmtFinSF1      4.750186
22     BsmtFinSF2      4.490713
23        LotArea      0.429203
24      BsmtUnfSF     -3.003196
25    OpenPorchSF     -3.143833
26        MiscVal     -3.910092
27         MoSold     -5.280061
28    GarageYrBlt    -17.382247
29       PoolArea    -61.926903
30    LotFrontage    -62.224525
31         YrSold   -191.225833
32     MSSubClass   -209.840235
33       HalfBath  -1955.444800
34   BedroomAbvGr  -9486.849775
35   KitchenAbvGr -21475.023726

In [39]:
y_pred = model.predict(X_test)

In [40]:
cv_scores = cross_val_score(
    model,
    X,
    y,
    scoring="neg_root_mean_squared_error",
    cv=5
)


In [41]:
rmse_scores = -cv_scores

In [42]:
print("RMSE scores:", rmse_scores)
print("Mean RMSE:", rmse_scores.mean())
print("Std RMSE:", rmse_scores.std())

RMSE scores: [28620.36158484 34666.43805385 36210.18305243 30308.23966965
 49433.72737805]
Mean RMSE: 35847.789947765
Std RMSE: 7335.215810649953


## reports
Well CV tells me that the model is usually around ~38$ off the actual value. also it tells me that the results must be very noisy.  

# Handling the missing values

# todo : later on this should be done via pipeline  

options  
1) replace with none (for features like PoolQC(pool quality)) [Missing means NONE]
2) replace with median [Missing but should exist]
3) drop [Missing because we don’t care (yet)]

## % missing values

In [43]:
miss_perc = (df.isnull().mean()*100)
miss_perc = miss_perc[miss_perc > 0]
miss_perc.sort_values(ascending=False)

Series([], dtype: float64)

In [44]:
none_type = ['PoolQC', 'MiscFeature', 'Alley', 'Fence', 'MasVnrType', 'FireplaceQu', 'GarageCond', 'GarageType', 'GarageFinish', 'GarageQual', 'BsmtFinType2', 'BsmtExposure',
             'BsmtQual', 'BsmtCond', 'BsmtFinType1', 'GarageYrBlt', 'MasVnrArea']

In [45]:
# for LotFrontage since 17% of the data is alot, we dont drop and also none doesn't seem right...
median_type = ['LotFrontage']

In [46]:
# A really small portion of the data is missing, just replace with mode (the most frequent value for this feature)
mode_type = ['Electrical']

In [47]:
# filling categorical features with 'None' and numerical with 0
for cat_feature in none_type:
    if df[cat_feature].dtype in ['float64', 'int64']:
        df[cat_feature].fillna(0, inplace=True)
    else:
        df[cat_feature].fillna('None', inplace=True)

In [48]:
# filling with the median
for median_feature in median_type:
    df[median_feature].fillna(df[median_feature].median(), inplace=True)

In [49]:
# filling with the mode
for mode_feature in mode_type:
    df[mode_feature].fillna(df[mode_feature].mode()[0], inplace=True)

## % missing values
after cleaning

In [50]:
miss_perc = (df.isnull().mean()*100)
miss_perc = miss_perc[miss_perc > 0]
miss_perc.sort_values(ascending=False)

Series([], dtype: float64)

## Train another model  
train another simple linear regression model with the filled missing values

In [51]:
# Select only numerical columns
df_numeric = df.select_dtypes(include=['int64', 'float64'])

print(f"Original shape: {df.shape}")
print(f"Numeric shape: {df_numeric.shape}")
print(f"Clean shape: {df_clean.shape}")

Original shape: (1460, 81)
Numeric shape: (1460, 38)
Clean shape: (1121, 38)


In [52]:
X = df_numeric.drop(['SalePrice', 'Id'], axis=1)
y = df_numeric['SalePrice']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=41)

In [53]:
model.fit(X_train, y_train)

LinearRegression()

In [54]:
coef_df1 = pd.DataFrame({
    'Feature': X.columns,
    'Coefficient': model.coef_
})
coef_df1 = coef_df1.sort_values('Coefficient', ascending=False).reset_index(drop=True)
coef_df1

Feature   Coefficient
0     OverallQual  17772.812182
1      GarageCars  17678.350189
2    BsmtFullBath   8654.057462
3    TotRmsAbvGrd   5280.989913
4     OverallCond   4946.641792
5        FullBath   4918.066980
6      Fireplaces   4542.758402
7    BsmtHalfBath   1468.252276
8       YearBuilt    290.239684
9    YearRemodAdd    130.512869
10    ScreenPorch     59.920894
11     MasVnrArea     34.413552
12      GrLivArea     26.812604
13     WoodDeckSF     24.970081
14      3SsnPorch     23.126333
15       2ndFlrSF     18.527093
16       1stFlrSF     17.648460
17  EnclosedPorch     13.456156
18     BsmtFinSF1     10.602820
19    TotalBsmtSF      8.038273
20     GarageArea      1.712126
21        LotArea      0.315776
22        MiscVal     -0.593975
23     BsmtFinSF2     -0.948386
24      BsmtUnfSF     -1.616161
25   LowQualFinSF     -9.362949
26    OpenPorchSF     -9.818491
27    GarageYrBlt    -14.644537
28       PoolArea    -32.817826
29    LotFrontage    -85.307885
30     MSSubClass   -204.512186
31         MoSold   -498.512268
32         YrSold  -1018.686052
33       HalfBath  -1251.805213
34   BedroomAbvGr -10834.144662
35   KitchenAbvGr -11638.812113

# Coefficient comparision

In [55]:
# FIX: your feature names are NOT the index — they are a column
# This version forces feature-name alignment and produces sane output

import pandas as pd

# assume first column = feature names, second column = coefficients
c0 = (
    coef_df0
    .iloc[:, :2]
    .set_index(coef_df0.columns[0])
    .iloc[:, 0]
    .pipe(pd.to_numeric, errors="coerce")
    .rename("model_0")
)

c1 = (
    coef_df1
    .iloc[:, :2]
    .set_index(coef_df1.columns[0])
    .iloc[:, 0]
    .pipe(pd.to_numeric, errors="coerce")
    .rename("model_1")
)

coef_compare = (
    pd.concat([c0, c1], axis=1, join="outer")
    .assign(
        abs_diff=lambda df: (df["model_1"] - df["model_0"]).abs(),
        sign_flip=lambda df: (df["model_0"] * df["model_1"]) < 0,
        present_in_both=lambda df: df["model_0"].notna() & df["model_1"].notna()
    )
    .sort_values("model_0", ascending=False)
)

coef_compare


model_0       model_1     abs_diff  sign_flip  \
Feature                                                             
OverallQual    19654.291953  17772.812182  1881.479771      False   
GarageCars     18344.552805  17678.350189   666.202616      False   
BsmtFullBath   11565.128964   8654.057462  2911.071502      False   
OverallCond     5072.289462   4946.641792   125.647670      False   
FullBath        5010.143758   4918.066980    92.076778      False   
Fireplaces      4793.918236   4542.758402   251.159835      False   
TotRmsAbvGrd    4587.157867   5280.989913   693.832046      False   
BsmtHalfBath    1338.748402   1468.252276   129.503874      False   
YearBuilt        286.898275    290.239684     3.341409      False   
YearRemodAdd     125.155507    130.512869     5.357362      False   
ScreenPorch       66.973618     59.920894     7.052724      False   
3SsnPorch         58.657103     23.126333    35.530770      False   
GrLivArea         29.344522     26.812604     2.531919      False   
MasVnrArea        26.492720     34.413552     7.920832      False   
WoodDeckSF        20.256803     24.970081     4.713279      False   
2ndFlrSF          13.903081     18.527093     4.624013      False   
EnclosedPorch     11.847823     13.456156     1.608332      False   
1stFlrSF          10.350316     17.648460     7.298143      False   
TotalBsmtSF        6.237704      8.038273     1.800569      False   
GarageArea         5.860430      1.712126     4.148305      False   
LowQualFinSF       5.091125     -9.362949    14.454074       True   
BsmtFinSF1         4.750186     10.602820     5.852633      False   
BsmtFinSF2         4.490713     -0.948386     5.439100       True   
LotArea            0.429203      0.315776     0.113427      False   
BsmtUnfSF         -3.003196     -1.616161     1.387035      False   
OpenPorchSF       -3.143833     -9.818491     6.674658      False   
MiscVal           -3.910092     -0.593975     3.316117      False   
MoSold            -5.280061   -498.512268   493.232207      False   
GarageYrBlt      -17.382247    -14.644537     2.737709      False   
PoolArea         -61.926903    -32.817826    29.109077      False   
LotFrontage      -62.224525    -85.307885    23.083360      False   
YrSold          -191.225833  -1018.686052   827.460219      False   
MSSubClass      -209.840235   -204.512186     5.328049      False   
HalfBath       -1955.444800  -1251.805213   703.639587      False   
BedroomAbvGr   -9486.849775 -10834.144662  1347.294887      False   
KitchenAbvGr  -21475.023726 -11638.812113  9836.211613      False   

               present_in_both  
Feature                         
OverallQual               True  
GarageCars                True  
BsmtFullBath              True  
OverallCond               True  
FullBath                  True  
Fireplaces                True  
TotRmsAbvGrd              True  
BsmtHalfBath              True  
YearBuilt                 True  
YearRemodAdd              True  
ScreenPorch               True  
3SsnPorch                 True  
GrLivArea                 True  
MasVnrArea                True  
WoodDeckSF                True  
2ndFlrSF                  True  
EnclosedPorch             True  
1stFlrSF                  True  
TotalBsmtSF               True  
GarageArea                True  
LowQualFinSF              True  
BsmtFinSF1                True  
BsmtFinSF2                True  
LotArea                   True  
BsmtUnfSF                 True  
OpenPorchSF               True  
MiscVal                   True  
MoSold                    True  
GarageYrBlt               True  
PoolArea                  True  
LotFrontage               True  
YrSold                    True  
MSSubClass                True  
HalfBath                  True  
BedroomAbvGr              True  
KitchenAbvGr              True

In [56]:
# applying CV
cv_scores = cross_val_score(
    model,
    X,
    y,
    scoring="neg_root_mean_squared_error",
    cv=5
)

In [57]:
rmse_scores = -cv_scores

In [58]:
print("RMSE scores:", rmse_scores)
print("Mean RMSE:", rmse_scores.mean())
print("Std RMSE:", rmse_scores.std())

RMSE scores: [28620.36158484 34666.43805385 36210.18305243 30308.23966965
 49433.72737805]
Mean RMSE: 35847.789947765
Std RMSE: 7335.215810649953


# comparision with the baseline 
## Baseline model results:  
RMSE scores: [29587.07709133 37665.99103449 40100.65600937 32378.41339688
 54086.83339133]  
Mean RMSE: 38763.79418468276  
Std RMSE: 8517.900283988969  

slight progress...
maybe less noise in folds
